src/uwtools/config/core.py

```diff
@@ -755,19 +682,153 @@ def print_config_section(config: dict, key_path: List[str]) -> None:
     key=value pairs, one per line.
     """
     keys = []
+    current_path = "<unknown>"
     for section in key_path:
         keys.append(section)
         current_path = " -> ".join(keys)
         try:
             subconfig = config[section]
-        except KeyError:
-            _log_and_error(f"Bad config path: {current_path}")
+        except KeyError as e:
+            raise _log_and_error(f"Bad config path: {current_path}") from e
         if not isinstance(subconfig, dict):
-            _log_and_error(f"Value at {current_path} must be a dictionary")
+            raise _log_and_error(f"Value at {current_path} must be a dictionary")
         config = subconfig
     output_lines = []
     for key, value in config.items():
         if type(value) not in (bool, float, int, str):
-            _log_and_error(f"Non-scalar value {value} found at {current_path}")
+            raise _log_and_error(f"Non-scalar value {value} found at {current_path}")
         output_lines.append(f"{key}={value}")
     print("\n".join(sorted(output_lines)))
+
+
+def realize_config(
+    input_file: OptionalPath,
+    input_format: str,
+    output_file: OptionalPath,
+    output_format: str,
+    values_file: OptionalPath,
+    values_format: Optional[str],
+    values_needed: bool = False,
+    dry_run: bool = False,
+) -> bool:
+    """
+    Realize an output config based on an input config and an optional values-providing config.
+
+    :param input_file: Input config file (stdin used when None).
+    :param input_format: Format of the input config.
+    :param output_file: Output config file (stdout used when None).
+    :param output_format: Format of the output config.
+    :param values_file: File providing values to modify input.
+    :param values_format: Format of the values config file.
+    :param values_needed: Provide a report about complete, missing, and template values.
+    :param dry_run: Log output instead of writing to output.
+    :return: True if no exception is raised.
+    :raises: UWConfigError if errors are encountered.
+    """
+
+    input_obj = _cli_name_to_config(input_format)(config_path=input_file)
+    input_obj.dereference_all()
+    input_obj = _realize_config_update(input_obj, values_file, values_format)
+    _realize_config_check_depths(input_obj, output_format)
+    if values_needed:
+        return _realize_config_values_needed(input_obj)
+    if dry_run:
+        logging.info(input_obj)
+    else:
+        _cli_name_to_config(output_format).dump_dict(path=output_file, cfg=input_obj.data)
+    return True
+
+
+# Private functions
+
+
+def _cli_name_to_config(cli_name: str) -> Type[Config]:
```

Comment:
The CLI expects (and enforces) certain values when referring to config formats. This function maps each identifier to the class that should be instantiated to deal with a config in that format.